qualtran/bloqs/arithmetic.py

```diff
@@ -208,7 +215,121 @@ def pretty_name(self) -> str:
+    Args:
+        bitsize: bitsize of x register.
+        val: integer to compare x against (a above.)
+
+    Registers:
+     - x: Registers to compare against val.
+     - z: Register to hold result of comparison.
+    """
+
+    bitsize: int
+    val: int
+
+    @cached_property
+    def signature(self) -> Signature:
+        return Signature.build(x=self.bitsize, z=1)
+
+    def t_complexity(self):
+        return t_complexity(LessThanGate(self.bitsize, self.val))
+
+    def bloq_counts(
+        self, ssa: Optional['SympySymbolAllocator'] = None
+    ) -> Set[Tuple[Union[int, sympy.Expr], Bloq]]:
+        tcomp = self.t_complexity()
+        return {(tcomp.t, TGate()), (tcomp.clifford, ArbitraryClifford(n=1))}
+
+
+@frozen
+class EqualsAConstant(Bloq):
+    r"""Implements $U_a|x\rangle = U_a|x\rangle|z\rangle = |x\rangle |z ^ (x == a)\rangle"
+
+    Args:
+        bitsize: bitsize of x register.
+        val: integer to compare x against (a above.)
+
+    Registers:
+     - x: Registers to compare against val.
+     - z: Register to hold result of comparison.
+    """
+
+    bitsize: int
+    val: int
+
+    @cached_property
+    def signature(self) -> Signature:
+        return Signature.build(x=self.bitsize, z=1)
+
+    def t_complexity(self):
+        return t_complexity(LessThanGate(self.bitsize, self.val))
+
+    def bloq_counts(
+        self, ssa: Optional['SympySymbolAllocator'] = None
+    ) -> Set[Tuple[Union[int, sympy.Expr], Bloq]]:
+        tcomp = self.t_complexity()
+        return {(tcomp.t, TGate()), (tcomp.clifford, ArbitraryClifford(n=1))}
+
+
+@frozen
+class ToContiguousIndex(Bloq):
+    r"""Build a contiguous register s from mu and nu.
+
+    $$
+        s = \nu (\nu + 1) / 2 + \mu
+    $$
+
+    Assuming nu is zero indexed (in contrast to the THC paper which assumes 1, hence the slightly different formula).
+
+    Args:
+        bitsize: number of bits for mu and nu registers.
+        s_bitsize: Number of bits for contiguous register.
+
+    Registers
+     - mu, nu: input registers
+     - s: output contiguous register
+
+    References:
+        (Even more efficient quantum computations of chemistry through
+        tensor hypercontraction)[https://arxiv.org/pdf/2011.03494.pdf] Eq. 29.
+    """
+
+    bitsize: int
+    s_bitsize: int
+
+    @cached_property
+    def signature(self) -> Signature:
+        return Signature(
+            [
+                Register("mu", bitsize=self.bitsize),
+                Register("nu", bitsize=self.bitsize),
+                Register("s", bitsize=self.s_bitsize),
+            ]
+        )
+
+    def on_classical_vals(
+        self, mu: 'ClassicalValT', nu: 'ClassicalValT'
+    ) -> Dict[str, 'ClassicalValT']:
+        return {'mu': mu, 'nu': nu, 's': nu * (nu + 1) // 2 + mu}
+
+    def bloq_counts(
+        self, ssa: Optional['SympySymbolAllocator'] = None
+    ) -> Set[Tuple[Union[int, sympy.Expr], Bloq]]:
+        return {(4 * (self.bitsize**2 + self.bitsize - 1), TGate())}
```

Comment:
it looks like Toffoli is the more natural cost here

Reply:
I'll move this discussion to #381